### PR TITLE
Constraint bridge #30

### LINKED ISSUE #30: [RFC] Functions and Set bridges
Solvers often support a specific set of (function, set) constraints but other constraint can be transformed into these pairs.
Therefore, different wrappers usually need to develop the same transformation without sharing code. Having these transformation in addition to the rest makes the wrapper code significantly more complicated (even more if it want to return ListOfConstraints, ... accurately).
Sometimes, we would also like to add the support of a (function, set) pair to a solver since we know to transform them into other constraints but we do not want to patch every solver since it would take time and our new constraint may be application specific and not of broad interest.

I think that these uses cases could be easy to solve if the following macro existed:
```julia
@bridge Name SomeFunction SomeSet SomeBridge
```
The macro would create an instance bridge `Name <: AbstractSolverInstance` such that `Name(instance)` creates an instance that supports all constraints supported by `instance` plus `(SomeFunction, SomeSet)`.
The macro would basically create
```julia
struct Name{InstanceType}
    instance::InstanceType
    bridges::Vector{SomeBridge}
end
```

I have already created some bridges [in SDOI](https://github.com/JuliaOpt/SemidefiniteOptInterface.jl/blob/master/src/setbridges.jl) and it works quite well. The macro does not exist yet though so I had to complicate the wrapper interface to add the `bridges::Vector{SomeBridge}` inside the wrapper structure, ...
Once the macro is created, this would allow to make SDOI a lot simpler.

Some examples of useful bridges:
* `AbstractVectorFunction` in `PositiveSemidefiniteConeSquare`, see https://github.com/JuliaOpt/JuMP.jl/pull/1125
* `AbstractVectorFunction` in `LogDetCone`, i.e. `log det(X) <= t`
* `AbstractVectorFunction` in `DetRootnCone`, i.e. `det(X) <= t^n`

### PR DESCRIPTION
**Pull request by blegat**
Will be merged in 24h if there is no objection (please ask if you need more time to review).
See [here](https://github.com/JuliaOpt/SemidefiniteOptInterface.jl/pull/6) for an usage example.
Closes #30

### CODE REVIEW

**Comment by blegat** [comment]
Ready for review :)

**Comment by blegat** [comment]
Added docs and tests. I will merge in 24h if there is no objection (please ask if you need more time to review).

File: src/intervalbridge.jl [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+struct SplitIntervalBridge{T}
+    lower::CR{MOI.ScalarAffineFunction{T}, MOI.GreaterThan{T}}
+    upper::CR{MOI.ScalarAffineFunction{T}, MOI.LessThan{T}}
+end
+function SplitIntervalBridge(instance, f::MOI.ScalarAffineFunction, s::MOI.Interval)
+    lower = MOI.addconstraint!(instance, f, MOI.GreaterThan(s.lower))
+    upper = MOI.addconstraint!(instance, f, MOI.LessThan(s.upper))
+    SplitIntervalBridge(lower, upper)
+end
+MOI.get(b::SplitIntervalBridge, ::MOI.NumberOfConstraints) = 0
```
Comment (mlubin) [inline src/intervalbridge.jl]: Please document what `MOI.get(::SplitIntervalBridge,...)` means, given that `SplitIntervalBridge` is not part of the MOI type hierarchy.

File: src/intervalbridge.jl [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+MOI.get(b::SplitIntervalBridge{T}, ::MOI.NumberOfConstraints{MOI.ScalarAffineFunction{T}, MOI.LessThan{T}}) where T = 1
+MOI.get(b::SplitIntervalBridge{T}, ::MOI.NumberOfConstraints{MOI.ScalarAffineFunction{T}, MOI.GreaterThan{T}}) where T = 1
+MOI.get(b::SplitIntervalBridge, ::MOI.ListOfConstraintReferences{F, S}) where {F, S} = CR{F, S}[]
+MOI.get(b::SplitIntervalBridge{T}, ::MOI.ListOfConstraintReferences{MOI.ScalarAffineFunction{T}, MOI.GreaterThan{T}}) where {T} = [b.lower]
+MOI.get(b::SplitIntervalBridge{T}, ::MOI.ListOfConstraintReferences{MOI.ScalarAffineFunction{T}, MOI.LessThan{T}}) where {T} = [b.upper]
+
+# References
+MOI.candelete(instance::MOI.AbstractInstance, c::SplitIntervalBridge) = true
```
Comment (mlubin) [inline src/intervalbridge.jl]: What does this mean?

File: src/intervalbridge.jl [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+function MOI.delete!(instance::MOI.AbstractInstance, c::SplitIntervalBridge)
+    MOI.delete!(instance, c.lower)
+    MOI.delete!(instance, c.upper)
+end
+
+# Attributes
+MOI.canget(instance::MOI.AbstractInstance, a::MOI.ConstraintPrimal, c::SplitIntervalBridge) = true
```
Comment (mlubin) [inline src/intervalbridge.jl]: Why does `SplitIntervalBridge` have both constraint attributes and instance attributes?

**Comment by mlubin** [comment]
@ulfworsoe, could you help review this PR? These bridges will be very relevant to the Mosek interface.

File: src/intervalbridge.jl (hunk 3 above)
Comment (blegat) [reply src/intervalbridge.jl]: It can act both as a Instance and as a constraint reference. Just like a bridge, it has two sides.

**Comment by mlubin** [comment]
I'm wondering if the bridge concept should be fleshed out more thoroughly and moved into MOI, but it's okay for now if you'd like to experiment with it as is. What's still missing is a description of what an `AbstractBridge` means and how it can sometimes act as an instance and sometimes as a constraint reference. If I'm having trouble following then I imagine others will as well.

**Comment by blegat** [comment]
@mlubin Thanks for the feedbacks !
I merge now so that we can start experimenting, we can still refine it in the process.
I agree that moving `AbstractBridge` to MOI might be a good idea, I have opened a poll [here](https://github.com/JuliaOpt/MathOptInterface.jl/issues/173).

**Comment by ulfworsoe** [comment]
Sorry for commenting late.

The idea looks good, and I can see several use cases, but there are clearly some points i am missing. I will need a bit more time to understand what is going on.

**Pull request merged**